File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.py
import sys

input = sys.stdin.readline

n = int(input())
maps = [[int(i) for i in input().split()] for _ in range(n)]


def comibiH(n, k):
    li = []
    result = []

    def recur():
        if len(li) == k:
            result.append(li[:])
            return

        for i in range(n):
            li.append(i)
            recur()
            li.pop()

    recur()
    return result


def compact(li):
    comp = [c for c in li if c != 0]
    for i in range(len(comp) - 1):
        if comp[i] == comp[i + 1]:
            comp[i] *= 2
            comp[i + 1] = 0
    comp = [c for c in comp if c != 0]

    return comp + ([0] * (len(li) - len(comp)))


def flip(maps):
    return [[maps[r][c] for r in range(len(maps))] for c in range(len(maps[0]))]


def left(maps):
    return [compact(r) for r in maps]


def right(maps):
    return [compact(r[::-1])[::-1] for r in maps]


def up(maps):
    return flip(left(flip(maps)))


def down(maps):
    return flip(right(flip(maps)))


funcs = [up, right, down, left]


def main():
    maxv = 0
    for li in comibiH(4, 5):
        n_maps = [r[:] for r in maps]
        for i in li:
            n_maps = funcs[i](n_maps)

        maxv = max(maxv, max([max(r) for r in n_maps]))

    return maxv


print(main())

File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.js
const fs = require("fs");
let read = fs.readFileSync("dev/stdin").toString().trim();

let [n, ...maps] = read.split("\n");
n = Number(n);

maps = maps.map((s) => s.split(" ").map(Number));

const compact = (li) => {
  let comp = li.filter((s) => s != 0);
  for (let i = 0; i < comp.length - 1; i++) {
    if (comp[i] == comp[i + 1]) {
      comp[i] *= 2;
      comp[i + 1] = 0;
    }
  }
  comp = comp.filter((s) => s != 0);
  const rest = new Array(li.length - comp.length).fill(0);
  return [...comp, ...rest];
};

const left = (maps) => {
  return maps.map((r) => compact(r));
};

const right = (maps) => {
  return maps.map((r) => compact(r.reverse()).reverse());
};

const flip = (maps) => {
  new_maps = [];
  for (let c = 0; c < maps[0].length; c++) {
    new_maps.push([]);
    for (let r = 0; r < maps.length; r++) {
      new_maps[c].push(maps[r][c]);
    }
  }
  return new_maps;
};

const up = (maps) => {
  return flip(left(flip(maps)));
};

const down = (maps) => {
  return flip(right(flip(maps)));
};

const funcs = [up, right, down, left];

const combiH = (n, k) => {
  const li = [];
  const result = [];
  const recur = () => {
    if (li.length == k) {
      result.push([...li]);
      return;
    }

    for (let i = 0; i < n; i++) {
      li.push(i);
      recur();
      li.pop();
    }
  };

  recur();

  return result;
};

const main = () => {
  let maxv = 0;
  for (let li of combiH(4, 5)) {
    let n_maps = maps.map((s) => Array.from(s));
    for (let i of li) {
      n_maps = funcs[i](n_maps);
    }

    maxv = Math.max(maxv, Math.max(...n_maps.map((s) => Math.max(...s))));
  }

  return maxv;
};

console.log(main());
